samples/augment-reality-to-show-hidden-infrastructure/src/main/java/com/esri/arcgismaps/sample/augmentrealitytoshowhiddeninfrastructure/components/AugmentedRealityViewModel.kt

```diff
@@ -0,0 +1,162 @@
+    )
+
+    init {
+        // For each pipe in the shared repository
+        SharedRepository.pipeInfoList.forEach {
+            viewModelScope.launch {
+                // Add Z values to the polyline using the base surface elevation
+                val polylineWithZ = densifyAndAddZValues(it)
+                // Add the 3D pipe to the pipe graphics overlay
+                pipeGraphicsOverlay.graphics.add(Graphic(polylineWithZ, pipeSymbol))
+                // Only add the shadow if the pipe is underground
+                if (it.elevationOffset < 0) {
+                    // Add the 2D pipe shadow to the shadow graphics overlay
+                    pipeShadowGraphicsOverlay.graphics.add(Graphic(it.polyline, pipeShadowSymbol))
+
+                    // Add leader lines connecting pipe vertices to shadow vertices
+                    addLeaderLines(polylineWithZ, it.elevationOffset)
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds Z values to the geometry by getting the elevation from the base surface.
+     */
+    private suspend fun densifyAndAddZValues(pipeInfo: PipeInfo): Polyline {
+        // Densify the polyline to ensure it has enough points for elevation sampling
+        val densifiedPolyline = GeometryEngine.densifyGeodeticOrNull(
+            geometry = pipeInfo.polyline,
+            maxSegmentLength = 1.0,
+            lengthUnit = LinearUnit.meters,
+            curveType = GeodeticCurveType.Geodesic
+        ) as Polyline
+        // Create a new polyline builder to construct the polyline with Z values
+        val polylineBuilder = PolylineBuilder(SpatialReference(3857))
+        // For each point in each part of the densified polyline
+        densifiedPolyline.parts.forEach { part ->
+            part.points.forEach { point ->
+                arcGISScene.baseSurface.elevationSources.first().load().onSuccess {
+                    arcGISScene.baseSurface.getElevation(point).let { elevationResult ->
+                        // Get the elevation at the point
+                        elevationResult.getOrNull()?.let { elevation ->
+                            // Add the point with the elevation offset to the polyline builder
+                            polylineBuilder.addPoint(
+                                GeometryEngine.createWithZ(
+                                    point,
+                                    elevation + pipeInfo.elevationOffset
+                                )
+                            )
+                        }
+                    }
+                }
+            }
+        }
+        return polylineBuilder.toGeometry()
+    }
+
+    /**
+     * Adds leader lines from the pipe vertices to the shadow vertices.
+     */
+    private fun addLeaderLines(pipePolyline: Polyline, elevationOffset: Float) {
+        // For each point in each part of the densified polyline
+        pipePolyline.parts.forEach { part ->
+            part.points.forEach { point ->
```

Comment:
Could you help explain why does there seem to be many more pipe vertices and leader lines than drawn in the initial screen? For example, I draw a pipe connecting 3 different points (in MapScreen), the AR view there are a lot more "squiggles" in the resulting polyline with many more leader lines:  

<img src="https://github.com/user-attachments/assets/64d2628a-ec5c-49e8-916f-8c6d82a3d211" height="500" />